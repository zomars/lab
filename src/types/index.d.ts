declare module 'typography-theme-wikipedia'
