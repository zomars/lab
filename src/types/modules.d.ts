declare module '*.woff2';
declare module '*.svg';
