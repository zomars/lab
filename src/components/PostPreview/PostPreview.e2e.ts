import { ElementHandle } from 'playwright';

import {
  ComponentWrapper,
  IComponentWrapperArgs,
} from '../../../e2e-test/components/component-wrapper.e2e';

enum PostPreviewSelector {
  host = '[data-testid*=PostPreview]',
  hostWithImage = '[data-testid*=PostPreview_withImage]',
  image = 'data-testid=PostPreview-Image',
  excerpt = 'data-testid=PostPreview-Excerpt',
  readMore = 'data-testid=PostPreview-ReadMore',
  header = 'data-testid=PostPreview-Header',
  tag = '[data-testid*=PostTags-Tag]',
}

interface IPostPreviewArgs extends Partial<IComponentWrapperArgs> {
  withPreview?: boolean;
}

export class PostPreview extends ComponentWrapper {
  constructor(args: IPostPreviewArgs = {}) {
    super({
      hostSelector: args.withPreview ? PostPreviewSelector.hostWithImage : PostPreviewSelector.host,
      ...args,
    });
  }

  public static async isTagLink($tag: ElementHandle): Promise<boolean> {
    const dataId = await $tag.getAttribute('data-testid');
    const dataIds = dataId.split(' ');

    return dataIds.includes('PostTags-Tag_link');
  }

  public getImage(): Promise<ElementHandle | null> {
    return this.getElementHandle(PostPreviewSelector.image);
  }

  public async getExcerptText(): Promise<string> {
    const $excerpt = await this.getElementHandle(PostPreviewSelector.excerpt);
    const text = await $excerpt.innerText();

    return text.trim();
  }

  public getTitle(): Promise<ElementHandle> {
    return this.getElementHandle(PostPreviewSelector.header);
  }

  public async getTitleText(): Promise<string> {
    const $title = await this.getTitle();
    const title = await $title.innerText();

    return title.trim();
  }

  public async getTags(): Promise<string[]> {
    const $elements = await this.getTagElements();

    const tags = $elements.map($element => $element.innerText());

    return Promise.all(tags)
      .then((tags: string[]) => tags.map(tag => tag.trim()));
  }

  public getTagElements(): Promise<ElementHandle[]> {
    return this.getElements(PostPreviewSelector.tag);
  }
}
