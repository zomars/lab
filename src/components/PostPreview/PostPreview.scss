@use '../../styles/variables' as vars;

.PostPreview {
  margin-bottom: var(--aml-global-layout-padding-md);
  overflow: hidden;

  &-Header {
    margin-top: 0;
  }

  &-Details {
    color: var(--aml-color-muted-text);
    line-height: var(--aml-line-height-none);
  }

  &-PreviewWrapper {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
  }

  &-ImageWrapper {
    display: flex;
    flex: 0 1 33%;
    align-items: center;
    justify-content: center;
  }

  &-Image {
    border: 2px solid var(--aml-color-gray-300);
    border-radius: var(--aml-radii-lg);
  }

  &-DetailsAndExcerpt {
    flex: 1 1 64%;
    margin-right: var(--aml-global-layout-padding-md);
  }

  &-ReadMore {
    flex-basis: 100%;
  }
}

@media (max-width: vars.$breakpoint-sm) {
  .PostPreview {
    &-ImageWrapper {
      flex-basis: 100%;
      margin-bottom: var(--aml-size-sm);
    }

    &-ReadMore {
      text-align: right;
    }
  }
}
