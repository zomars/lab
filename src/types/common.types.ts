import { Node } from 'gatsby';
import { ReactNode } from 'react';
import { ImageDataLike } from 'gatsby-plugin-image';

export interface INodeFields {
  [key: string]: unknown,
}

interface INode extends Node {
  fields: INodeFields;
}

interface IBlogPostFrontmatter {
  title: string,
  date: string,
  tags: string[],
  coverImage?: {
    childImageSharp: ImageDataLike,
  },
  published?: boolean,
  updated?: string,
  summary?: string & ReactNode,
  description: string,
}

interface IReadingTime {
  text: string,
  words: number,
}

interface IBlogPostFields extends INodeFields {
  slug: string,
  readingTime: IReadingTime,
}

export interface IBlogPost extends INode {
  frontmatter: IBlogPostFrontmatter;
  fields: IBlogPostFields;
  excerpt?: string;
  body?: string;
}

export interface IUniquePostTag {
  name: string;
  count: number;
}

export interface ISiteMetadata {
  title: string;
  description: string;
  author: string;
  siteUrl: string;
}

export interface IReactNodeProps {
  children: string | ReactNode;
}

export interface IPostTagsMappingQResponse {
  allPosts: {
    posts: IBlogPost[],
  },
  uniqueTags: {
    tags: IUniquePostTag[],
  },
}
