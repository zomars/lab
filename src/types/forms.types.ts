export interface ISendMessageFormPayload {
  message: string;
  captcha: string;
}
