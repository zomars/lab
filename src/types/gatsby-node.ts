import { Node } from 'gatsby';

export interface INodeFields {
  slug: string,
  [key: string]: unknown,
}

export interface INode extends Node {
  fields: INodeFields;
}
