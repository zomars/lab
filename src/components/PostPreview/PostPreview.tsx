import {
  GatsbyImage,
  getImage,
  IGatsbyImageData,
} from 'gatsby-plugin-image';
import { MDXRenderer } from 'gatsby-plugin-mdx';
import React, { ReactElement } from 'react';
import { Link } from 'gatsby';
import { cn } from '@bem-react/classname';
import { classnames } from '@bem-react/classnames';

import { IBlogPost } from '../../types/common.types';

import './PostPreview.scss';

const cnPostPreview = cn('PostPreview');

/**
 * Return post date, reading time and list of tags.
 */
function getDetailsString(post: IBlogPost): string {
  const { date: posted, updated } = post.frontmatter;
  const { text } = post.fields.readingTime;

  let date = posted;

  if (updated) {
    date += ` (updated ${ updated })`;
  }

  const details = [
    date,
    text,
    // `${words} words`,
  ];

  return details.join(' â€” ');
}

export function PostPreview(props: {
  className?: string;
  testId?: string;
  post: IBlogPost;
  tag: string;
}): ReactElement {
  const {
    post,
    className,
    tag,
  } = props;

  const { slug } = post.fields;
  const { title, coverImage, summary } = post.frontmatter;

  // so that we have active tag info in the runtime
  // for the PostPage prev/next links
  const blogPostLinkPayload = {
    activeTag: tag,
  };

  let CoverImage = null;

  if (coverImage) {
    const imageData = getImage(coverImage.childImageSharp) as IGatsbyImageData;

    CoverImage = (
      <Link
        className = { cnPostPreview('ImageWrapper') }
        to = { slug }
        state = { blogPostLinkPayload }
      >
        <GatsbyImage
          data-testid = { cnPostPreview('Image') }
          imgClassName = { cnPostPreview('Image') }
          alt = { `Cover photo for ${ title } post` }
          loading = 'eager'
          image = { imageData }
        />
      </Link>
    );
  }

  const { testId } = props;

  return (
    <div
      className = { classnames(cnPostPreview(), className) }
      data-testid = {
        classnames(
          cnPostPreview({ withImage: !!coverImage }),
          testId,
        )
      }
    >
      <h3 className = { cnPostPreview('Header') }>
        <Link
          data-testid = { cnPostPreview('Header') }
          to = { slug }
          state = { blogPostLinkPayload }
        >
          { title }
        </Link>
      </h3>

      <div className = { cnPostPreview('PreviewWrapper') }>
        <p
          className = { cnPostPreview('Details') }
          data-testid = { cnPostPreview('Details') }
        >
          { getDetailsString(post) }
        </p>

        <Link
          data-testid = { cnPostPreview('Excerpt') }
          className = { cnPostPreview('Excerpt') }
          to = { slug }
          state = { blogPostLinkPayload }
        >
          { summary ? <MDXRenderer>{ summary }</MDXRenderer> : <p> { post.excerpt! } </p>}
        </Link>

        { CoverImage }
      </div>
    </div>
  );
}
