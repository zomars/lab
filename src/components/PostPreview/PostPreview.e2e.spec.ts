import { IGlobal } from '../../../e2e-test/e2e.types';
import { PostPreview } from './PostPreview.e2e';

const localGlobal = global as IGlobal & typeof globalThis;

describe('site header', () => {
  beforeEach(async () => {
    await jestPlaywright.resetContext();

    context.setDefaultTimeout(localGlobal.defaultTimeout);

    await page.goto(`${ localGlobal.url }/blog/cars/1`);
  });

  it('random (first) preview has all expected elements', async () => {
    const postPreview = new PostPreview();

    await expect(postPreview.isConnected).resolves.toBe(true);

    await expect(postPreview.getTitleText()).resolves.toEqual(expect.any(String));

    await expect(postPreview.getExcerptText()).resolves.toEqual(expect.any(String));

    await expect(postPreview.getReadMore()).resolves.toBeTruthy();
  });

  it('post with image is present and has an image', async () => {
    const postPreview = new PostPreview({ withPreview: true });

    await expect(postPreview.isConnected).resolves.toBe(true);

    await expect(postPreview.getImage()).resolves.toBeTruthy();
  });

  // @todo need to handle links open in other tabs
  it.skip('has three areas to click in order to get to the post', async () => {
    const postPreview = new PostPreview({ withPreview: true });

    await expect(postPreview.isConnected).resolves.toBe(true);

    const $title = await postPreview.getTitle();

    $title.click()
      .then();
  });
});
