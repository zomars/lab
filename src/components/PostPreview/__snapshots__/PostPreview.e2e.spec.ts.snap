// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`site header first preview has all expected elements 1`] = `"Invalidate or Clear Github Actions Cache"`;

exports[`site header first preview has all expected elements 2`] = `"Cache invalidation is one of the most complex computer science tasks. Especially so for GitHub action workflows since there is no embedded mechanism to clean or invalidate present caches. However, what we can do is to update cache keys with a dedicated environment variable of the workflow or repository secret."`;

exports[`site header first preview has all expected elements 3`] = `"May 13, 2022"`;

exports[`site header first preview has all expected elements 4`] = `"4 min read"`;

exports[`site header post link click works on excerpt 1`] = `"/posts/2021/dependabot-pull-request-github-actions/"`;

exports[`site header post link click works on image 1`] = `"/posts/2021/dependabot-pull-request-github-actions/"`;

exports[`site header post link click works on title 1`] = `"/posts/2021/dependabot-pull-request-github-actions/"`;
